src/tests/ftest/util/apricot/apricot/test.py

```diff
@@ -1242,8 +1248,33 @@ def remove_temp_test_dir(self):
             errors.append("Error removing temporary test files")
         return errors
 
+    def fail(self, message=None):
+        if self.dumped_engines_stacks is False:
+            self.dumped_engines_stacks = True
+            self.log.info("Test status has failed, dumping ULT stacks")
+            dump_engines_stacks(self.hostlist_servers)
+        super().fail(message)
+
+    def error(self, message=None):
+        if self.dumped_engines_stacks is False:
+            self.dumped_engines_stacks = True
+            self.log.info("Test status has errored, dumping ULT stacks")
+            dump_engines_stacks(self.hostlist_servers)
+        super().error(message)
+
     def tearDown(self):
         """Tear down after each test case."""
+
+        # dump engines ULT stacks upon test failure
+        # XXX check of Avocado test status during teardown is presently useless
```

Comment:
(pylint-fixme) XXX check of Avocado test status during teardown is presently useless